File: app/src/main/java/com/ywwynm/everythingdone/model/ThingAction.java
package com.ywwynm.everythingdone.model;

import android.os.Bundle;
import android.support.annotation.IntDef;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by ywwynm on 2016/7/1.
 * Thing update action model
 */
public class ThingAction {

    public static final int UPDATE_TITLE             = 0;
    public static final int UPDATE_CONTENT           = 1;
    public static final int TOGGLE_CHECKLIST         = 2;
    public static final int UPDATE_CHECKLIST         = 3;
    public static final int MOVE_CHECKLIST           = 4;
    public static final int UPDATE_COLOR             = 5;
    public static final int ADD_ATTACHMENT           = 6;
    public static final int DELETE_ATTACHMENT        = 7;
    public static final int MOVE_ATTACHMENT          = 8;
    public static final int TOGGLE_REMINDER_OR_HABIT = 9;
    public static final int UPDATE_REMINDER_OR_HABIT = 10;
    public static final int TOGGLE_PRIVATE           = 11;

    @IntDef({UPDATE_TITLE, UPDATE_CONTENT,
            TOGGLE_CHECKLIST, UPDATE_CHECKLIST, MOVE_CHECKLIST,
            UPDATE_COLOR,
            ADD_ATTACHMENT, DELETE_ATTACHMENT, MOVE_ATTACHMENT,
            TOGGLE_REMINDER_OR_HABIT, UPDATE_REMINDER_OR_HABIT,
            TOGGLE_PRIVATE})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Type {}

    public static final String KEY_ATTACHMENT_TYPE   = "attachment_type";
    public static final String KEY_CHECKBOX_STATE    = "checkbox_state";
    public static final String KEY_CURSOR_POS_BEFORE = "cursor_pos_before";
    public static final String KEY_CURSOR_POS_AFTER  = "cursor_pos_after";
    public static final String KEY_PICKED_BEFORE     = "picked_before";
    public static final String KEY_PICKED_AFTER      = "picked_after";

    private @Type int mType;

    private Object mBefore;

    private Object mAfter;

    private Bundle mExtras;

    public ThingAction(@Type int type, Object before, Object after) {
        mType   = type;
        mBefore = before;
        mAfter  = after;
        mExtras = new Bundle();
    }

    public int getType() {
        return mType;
    }

    public Object getBefore() {
        return mBefore;
    }

    public Object getAfter() {
        return mAfter;
    }

    public Bundle getExtras() {
        return mExtras;
    }
}

File: app/src/main/java/com/ywwynm/everythingdone/model/ThingAction.kt
package com.ywwynm.everythingdone.model

import android.os.Bundle
import androidx.annotation.IntDef

/**
 * Created by ywwynm on 2016/7/1.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * Thing update action model
 */
open class ThingAction(private val type: Int, private val before: Any?, private val after: Any?) {

    private val extras: Bundle = Bundle()

    open fun getType(): Int {
        return type
    }

    open fun getBefore(): Any? {
        return before
    }

    open fun getAfter(): Any? {
        return after
    }

    open fun getExtras(): Bundle? {
        return extras
    }

    @IntDef(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13)
    @Retention(AnnotationRetention.SOURCE)
    annotation class Type

    companion object {
        const val UPDATE_TITLE: Int             = 0
        const val UPDATE_CONTENT: Int           = 1
        const val TOGGLE_CHECKLIST: Int         = 2
        const val UPDATE_CHECKLIST: Int         = 3
        const val MOVE_CHECKLIST: Int           = 4
        const val UPDATE_COLOR: Int             = 5
        const val ADD_ATTACHMENT: Int           = 6
        const val DELETE_ATTACHMENT: Int        = 7
        const val MOVE_ATTACHMENT: Int          = 8
        const val TOGGLE_REMINDER_OR_HABIT: Int = 9
        const val UPDATE_REMINDER_OR_HABIT: Int = 10
        const val TOGGLE_PRIVATE: Int           = 11
        const val UPDATE_HOME_CARD_SPAN_MODE: Int = 12
        const val UPDATE_HOME_CARD_IMAGE_PLACEMENT: Int = 13

        const val KEY_ATTACHMENT_TYPE: String   = "attachment_type"
        const val KEY_CHECKBOX_STATE: String    = "checkbox_state"
        const val KEY_CURSOR_POS_BEFORE: String = "cursor_pos_before"
        const val KEY_CURSOR_POS_AFTER: String  = "cursor_pos_after"
        const val KEY_PICKED_BEFORE: String     = "picked_before"
        const val KEY_PICKED_AFTER: String      = "picked_after"
    }
}
